BasicHistoryApi/app/src/main/java/com/google/android/gms/fit/samples/basichistoryapi/MainActivity.kt

```diff
@@ -0,0 +1,431 @@
+                .addDataType(DataType.AGGREGATE_STEP_COUNT_DELTA, FitnessOptions.ACCESS_WRITE)
+                .build()
+    }
+
+    // Two maps are used in conjunction with the OAuth flow. This is in order to identify what
+    // function to execute once the flow has successfully completed.
+    private val signInActionMap = mapOf<() -> Any, Int>(
+            ::insertAndReadData to 1,
+            ::updateAndReadData to 2,
+            ::deleteData to 3
+    )
+    private val signInResponseMap = signInActionMap.entries.associate { (k, v) -> v to k }
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_main)
+
+        initializeLogging()
+
+        fitSignIn(::insertAndReadData)
+    }
+
+    /**
+     * Checks that the user is signed in, and if so, executes the specified function. If the user is
+     * not signed in, initiates the sign in flow, specifying the post-sign in function to execute.
+     *
+     * @param signInAction The function to execute if already signed in, or once signed in.
+     */
+    private fun fitSignIn(signInAction: () -> Any) {
+        if (!GoogleSignIn.hasPermissions(getGoogleAccount(), fitnessOptions)) {
+            val requestCode = signInActionMap[signInAction]
+            requestCode?.let {
+                GoogleSignIn.requestPermissions(
+                        this,
+                        requestCode,
+                        getGoogleAccount(), fitnessOptions)
+            }
+        } else signInAction()
+    }
+
+    private fun fitSignOut() {
+        val signInOptions = GoogleSignInOptions.Builder().addExtension(fitnessOptions).build()
+        val client = GoogleSignIn.getClient(this, signInOptions)
+        client.revokeAccess().addOnSuccessListener { Log.i(TAG, "Signed out") }
+    }
+
+    /**
+     * Handles the callback from the OAuth sign in flow, executing the post sign in function
+     */
+    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
+        super.onActivityResult(requestCode, resultCode, data)
+
+        when (resultCode) {
+            RESULT_OK -> {
+                val postSignInAction = signInResponseMap[requestCode]
+                postSignInAction?.let {
+                    postSignInAction()
+                }
+            }
+            else -> oAuthErrorMsg(requestCode, resultCode)
+        }
+    }
+
+    private fun oAuthErrorMsg(requestCode: Int, resultCode: Int) {
+        val message = """
+            There was an error signing into Fit. Check the troubleshooting section of the README
+            for potential issues.
+            Request code was: $requestCode
+            Result code was: $resultCode
+        """.trimIndent()
+        Log.e(TAG, message)
+    }
+
+    private fun getGoogleAccount() = GoogleSignIn.getAccountForExtension(this, fitnessOptions)
```

Comment:
I'm not familiar with this method (getAccountForExtension), can you add a short comment explaining what it does?